classes/local/store/object_file_system.php

```diff
@@ -401,6 +401,29 @@ public function get_content(\stored_file $file) {
         return $contents;
     }
 
+    /**
+     * Serve file content using X-Sendfile header.
+     * Please make sure that all headers are already sent and the all
+     * access control checks passed.
+     *
+     * This alternate method to xsendfile() allows an alternate file system
+     * to use the full file metadata and avoid extra lookups.
+     *
+     * @param stored_file $file The file to send
+     * @return bool success
+     * @throws \dml_exception
+     */
+    public function xsendfile_file(stored_file $file): bool {
+        $contenthash = $file->get_contenthash();
+        if ($this->presigned_url_configured() &&
+                $this->presigned_url_should_redirect_file($file) &&
+                $this->is_file_readable_externally_by_hash($contenthash)) {
```

Comment:
I would profile this, but I would expect a db call to be less overhead than an s3 call. It's probably very marginal.

But some moodles (not ours) will have objectfs remote, so I'd lean towards db first